src/org/sosy_lab/java_smt/delegate/debugging/DebuggingSolverContext.java

```diff
@@ -0,0 +1,239 @@
+// This file is part of JavaSMT,
+// an API wrapper for a collection of SMT solvers:
+// https://github.com/sosy-lab/java-smt
+//
+// SPDX-FileCopyrightText: 2024 Dirk Beyer <https://www.sosy-lab.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package org.sosy_lab.java_smt.delegate.debugging;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+import org.sosy_lab.common.configuration.Option;
+import org.sosy_lab.common.configuration.Options;
+import org.sosy_lab.java_smt.SolverContextFactory.Solvers;
+import org.sosy_lab.java_smt.api.Formula;
+import org.sosy_lab.java_smt.api.FormulaManager;
+import org.sosy_lab.java_smt.api.FunctionDeclaration;
+import org.sosy_lab.java_smt.api.FunctionDeclarationKind;
+import org.sosy_lab.java_smt.api.InterpolatingProverEnvironment;
+import org.sosy_lab.java_smt.api.OptimizationProverEnvironment;
+import org.sosy_lab.java_smt.api.ProverEnvironment;
+import org.sosy_lab.java_smt.api.SolverContext;
+import org.sosy_lab.java_smt.api.visitors.DefaultFormulaVisitor;
+import org.sosy_lab.java_smt.api.visitors.TraversalProcess;
+
+// TODO: Figure out what to do about boolector (does not support visitors..)
+// TODO: Add an option that checks if sorts were created on the context
+
+@Options(prefix = "solver.debugMode")
+public class DebuggingSolverContext extends DefaultFormulaVisitor<TraversalProcess>
```

Comment:
This class mixes three contepts, which are SolverContext, Visitor, and a central point for DebugAssertions. It might be easier to use a seperated classes for those three concepts, i.e., a separated visitor and a specialized DebuggingContext that contains the checks. Those three classes can share fields if required.

Reply:
I've split up the class now. `DebuggingSolverContext` now only contains the methods of the JavaSMT interface, a new class `DebuggingContext`is used for the options and the bookkeeping data, and the visitor method was moved to an anonymous class that is created directly in `addFormulaTerm`. Hopefully this will make the code a little easier to read.